ttab/model_adaptation/rotta.py

```diff
@@ -0,0 +1,664 @@
+            saturation_factor = random.uniform(saturation[0], saturation[1])
+            transforms.append(
+                Lambda(lambda img: F.adjust_saturation(img, saturation_factor))
+            )
+
+        if hue is not None:
+            hue_factor = random.uniform(hue[0], hue[1])
+            transforms.append(Lambda(lambda img: F.adjust_hue(img, hue_factor)))
+
+        if gamma is not None:
+            gamma_factor = random.uniform(gamma[0], gamma[1])
+            transforms.append(Lambda(lambda img: F.adjust_gamma(img, gamma_factor)))
+
+        random.shuffle(transforms)
+        transform = Compose(transforms)
+
+        return transform
+
+    def forward(self, img):
+        """
+        Args:
+            img (PIL Image or Tensor): Input image.
+
+        Returns:
+            PIL Image or Tensor: Color jittered image.
+        """
+        fn_idx = torch.randperm(5)
+        for fn_id in fn_idx:
+            if fn_id == 0 and self.brightness is not None:
+                brightness = self.brightness
+                brightness_factor = (
+                    torch.tensor(1.0).uniform_(brightness[0], brightness[1]).item()
+                )
+                img = F.adjust_brightness(img, brightness_factor)
+
+            if fn_id == 1 and self.contrast is not None:
+                contrast = self.contrast
+                contrast_factor = (
+                    torch.tensor(1.0).uniform_(contrast[0], contrast[1]).item()
+                )
+                img = F.adjust_contrast(img, contrast_factor)
+
+            if fn_id == 2 and self.saturation is not None:
+                saturation = self.saturation
+                saturation_factor = (
+                    torch.tensor(1.0).uniform_(saturation[0], saturation[1]).item()
+                )
+                img = F.adjust_saturation(img, saturation_factor)
+
+            if fn_id == 3 and self.hue is not None:
+                hue = self.hue
+                hue_factor = torch.tensor(1.0).uniform_(hue[0], hue[1]).item()
+                img = F.adjust_hue(img, hue_factor)
+
+            if fn_id == 4 and self.gamma is not None:
+                gamma = self.gamma
+                gamma_factor = torch.tensor(1.0).uniform_(gamma[0], gamma[1]).item()
+                img = img.clamp(
+                    1e-8, 1.0
+                )  # to fix Nan values in gradients, which happens when applying gamma
+                # after contrast
+                img = F.adjust_gamma(img, gamma_factor)
+
+        return img
+
+    def __repr__(self):
+        format_string = self.__class__.__name__ + "("
+        format_string += "brightness={0}".format(self.brightness)
+        format_string += ", contrast={0}".format(self.contrast)
+        format_string += ", saturation={0}".format(self.saturation)
+        format_string += ", hue={0})".format(self.hue)
+        format_string += ", gamma={0})".format(self.gamma)
+        return format_string
+
+
+# ========= utils ==========
```

Comment:
Move irrelevant functions to `utils.py` or make it internal for TTA method class if you have to.